modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItAlterTableDdlTest.java

```diff
@@ -0,0 +1,223 @@
+
+        assertQuery("SELECT * FROM my")
+                .returns(1, -1)
+                .check();
+    }
+
+    /** Test that adding nullable column via ALTER TABLE ADD name type NULL works. */
+    @Test
+    public void testNullableColumn() {
+        sql("CREATE TABLE my (c1 INT PRIMARY KEY, c2 INT)");
+        sql("INSERT INTO my VALUES (1, 1)");
+        sql("ALTER TABLE my ADD COLUMN c3 INT NULL");
+        sql("INSERT INTO my VALUES (2, 2, NULL)");
+
+        assertQuery("SELECT * FROM my ORDER by c1 ASC")
+                .returns(1, 1, null)
+                .returns(2, 2, null)
+                .check();
+    }
+
+    /**
+     * Adds columns of all supported types and checks that the row
+     * created on the old schema version is read correctly.
+     */
+    @Test
+    public void testDropAndAddColumnsAllTypes() {
+        List<NativeType> allTypes = SchemaTestUtils.ALL_TYPES;
+
+        Set<NativeTypeSpec> unsupportedTypes = Set.of(
+                // TODO https://issues.apache.org/jira/browse/IGNITE-18431
+                NativeTypeSpec.BITMASK
+        );
+
+        // List of columns for 'ADD COLUMN' statement.
+        IgniteStringBuilder addColumnsList = new IgniteStringBuilder();
+        // List of columns for 'DROP COLUMN' statement.
+        IgniteStringBuilder dropColumnsList = new IgniteStringBuilder();
+
+        for (int i = 0; i < allTypes.size(); i++) {
+            NativeType type = allTypes.get(i);
+
+            if (unsupportedTypes.contains(type.spec())) {
+                continue;
+            }
+
+            RelDataType relDataType = TypeUtils.native2relationalType(Commons.typeFactory(), type);
+
+            if (addColumnsList.length() > 0) {
+                addColumnsList.app(',');
+                dropColumnsList.app(',');
+            }
+
+            addColumnsList.app("c").app(i).app(' ').app(relDataType.getSqlTypeName().getSpaceName());
+            dropColumnsList.app("c").app(i);
+        }
+
+        sql("CREATE TABLE test (id INT PRIMARY KEY, val INT)");
+        sql("INSERT INTO test VALUES (0, 1)");
+        sql(format("ALTER TABLE test ADD COLUMN ({})", addColumnsList.toString()));
+
+        List<List<Object>> res = sql("SELECT * FROM test");
+        assertThat(res.size(), is(1));
+        assertThat(res.get(0).size(), is(allTypes.size() - unsupportedTypes.size() + /* initial columns */ 2));
+
+        sql(format("ALTER TABLE test DROP COLUMN ({})", dropColumnsList.toString()));
+        assertQuery("SELECT * FROM test")
+                .returns(0, 1)
+                .check();
+    }
+
+
+    @Test
+    public void addColumnWithConstantDefault() {
+        sql("CREATE TABLE test(id BIGINT DEFAULT 1 PRIMARY KEY)");
+
+        sql("ALTER TABLE test ADD COLUMN valint INTEGER DEFAULT 1");
+        sql("ALTER TABLE test ADD COLUMN valdate DATE DEFAULT DATE '2001-12-21'");
+        sql("ALTER TABLE test ADD COLUMN valtime TIME(3) DEFAULT TIME '11:22:33.444555'");
```

Comment:
i see that`s is not your stuff, but it we change TIME(3) into TIME(4) test still will be passes, seems we need to link appropriate issue here. The same for TIMESTAMP.

Reply:
I've added a todo with a link to the ticke.